crates/bevy_ecs/src/bundle.rs

```diff
@@ -683,7 +691,12 @@ impl<'a, 'b> BundleSpawner<'a, 'b> {
 #[derive(Default)]
 pub struct Bundles {
     bundle_infos: Vec<BundleInfo>,
+    /// Cache static [`BundleId`]
     bundle_ids: HashMap<TypeId, BundleId>,
+    /// Cache dynamic [`BundleId`] with multiple components
+    dynamic_bundle_ids: HashMap<Vec<ComponentId>, (BundleId, Vec<StorageType>)>,
```

Comment:
Non-blocking: I feel like this could be better represented by putting `StorageType` back into `BundleInfo` and replacing the `Vec<ComponentId>` with `Vec<BundleComponent>` which contains both the ID and the storage type. The same for the below.

Reply:
Can definitely do this in a follow-up for removing dynamic components!